hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/ClusterJoinManager.java

```diff
@@ -309,22 +305,20 @@ private boolean checkClusterStateBeforeJoin(Address target, String uuid) {
             return checkRecentlyJoinedMemberUuidBeforeJoin(target, uuid);
         }
 
-        if (clusterService.isMemberRemovedInNotJoinableState(target)) {
-            MemberImpl removedMember = clusterService.getMembershipManager().getMemberRemovedInNotJoinableState(uuid);
-
-            if (removedMember != null && !target.equals(removedMember.getAddress())) {
-
-                logger.warning("UUID " + uuid + " was being used by " + removedMember
+        // RU_COMPAT_3_11
+        if (clusterService.getClusterVersion().isLessThan(Versions.V3_12) && clusterService.isMissingMember(target, uuid)) {
```

Comment:
In 3.11.x code:

- if a member returns with the same address but with _a UUID that matches to another address previously_, it is rejected
- if a member returns with the same UUID but _different address that doesn't match to another UUID previously_, it is accepted

So, essentially a member should restart using either the same `<UUID, address>` pair, or `<UUID, new-address>` pair or `<new-UUID, address>` pair. It cannot inherit UUID and address from two members. 

Without hot restart, a member cannot restore its former UUID, only it can recover its address (unless you dive into internals). So this check will be same for 3.11.x and 3.12.

With hot restart, for RU scenarios, I think behaviour will remain the same, as I described above.

When cluster version is 3.12, without hot restart; 
- a node restarting with the same address is accepted, , regardless of UUID (which will be a new UUID, since cannot be restored)
- a node restarting with a new address is rejected, regardless of UUID

When cluster version is 3.12, with hot restart, a node restarting with the same UUID is accepted, regardless of the address.

Reply:
I think we agree on both old (3.11.x) and new 3.12 behaviour, the only difference being in behaviour during RU (the second and third I've mentioned, introduced by this PR). So let me focus on that one. I might be wrong here so please let me know what I'm missing. The code under RU is basically this (if we assume `clusterService.getClusterVersion().isLessThan(Versions.V3_12) == true`):

```java
        if (clusterService.isMissingMember(target, uuid)) {
            MemberImpl missingMember = clusterService.getMembershipManager().getMissingMember(target, uuid);
            if (missingMember != null && !target.equals(missingMember.getAddress())) {
                logger.warning("UUID " + uuid + " was being used by " + missingMember
                        + " before. " + target + " is not allowed to join with a UUID which belongs to"
                        + " a known passive member.");

                return true;
            }
            return false;
        }

        if (clusterService.isMissingMember(target, uuid)) {
            return false;
        }
```

If we're using OS, we are comparing missing members based on address. 
- if the joining member has an address of a missing member, we will let it join in the new RU code, regardless of UUID (so it might be a completely new member on the same address). In 3.11.x code we might have rejected it if the UUID was different than the one of the missing member.
- if the joining member has a new address but an UUID of a missing member, we will reject it. In 3.11.x code we accepted such a member. This scenario probably isn't that important as we are running OS and it would assume the member has changed its address or managed to persist the UUID between restarts somehow.

If we're running EE with HR, we are comparing missing members based on UUID.
- if the joining member has an UUID of a missing member but a different address, the 3.11.x master would accept it while the new RU code would reject it.
- both 3.11.x and the new code reject instances with new UUIDs but with an address of a missing member. No difference here.

Another note is that we have two identical if's when running under RU (`if (clusterService.isMissingMember(target, uuid))`) but that's not that important.

Reply:
When using OS (or specifically without hot restart), a member cannot restore its UUID (at least we don't support it), so restarting member's UUID will be always a new one. That's why I think behaviour will be same for 3.11 and 3.12 under RU.

When using EE with HR;
> if the joining member has an UUID of a missing member but a different address, the 3.11.x master would accept it while the new RU code would reject it.

I think you're right here. I'll think how to solve this.

Reply:
@mmedenjak: Can you check the new version?

Reply:
Looks good, thanks!